internal/engine/procedures/typing/typing.go

```diff
@@ -0,0 +1,663 @@
+	return types.NullType
+}
+
+func (t *typingVisitor) VisitExpressionParenthesized(p0 *parser.ExpressionParenthesized) any {
+	return p0.Expression.Accept(t)
+}
+
+func (t *typingVisitor) VisitExpressionTextLiteral(p0 *parser.ExpressionTextLiteral) any {
+	return types.TextType
+}
+
+func (t *typingVisitor) VisitExpressionVariable(p0 *parser.ExpressionVariable) any {
+	dt, ok := t.declarations[p0.Name]
+	if !ok {
+		anonType, ok := t.anonymousDeclarations[p0.Name]
+		if !ok {
+			panic(fmt.Sprintf("variable %s not declared", p0.Name))
+		}
+
+		return anonType
+	}
+
+	return dt
+}
+
+func (t *typingVisitor) VisitLoopTargetCall(p0 *parser.LoopTargetCall) any {
+	r := t.analyzeProcedureCall(p0.Call)
+
+	if !r.IsTable {
+		panic("loops on procedures must return a table")
+	}
+
+	vals := make(map[string]*types.DataType)
+	for _, col := range r.Fields {
+		vals[col.Name] = col.Type
+	}
+
+	return vals
+}
+
+func (t *typingVisitor) VisitLoopTargetRange(p0 *parser.LoopTargetRange) any {
+	t.asserIntType(p0.Start)
+	t.asserIntType(p0.End)
+	return types.IntType
+}
+
+func (t *typingVisitor) VisitLoopTargetSQL(p0 *parser.LoopTargetSQL) any {
+	rel := t.analyzeSQL(p0.Statement)
+
+	r := make(map[string]*types.DataType)
+	err := rel.Loop(func(s string, a *engine.Attribute) error {
+		r[s] = a.Type
+		return nil
+	})
+	if err != nil {
+		panic(err)
+	}
+	return r
+}
+
+func (t *typingVisitor) VisitLoopTargetVariable(p0 *parser.LoopTargetVariable) any {
+	// must check that the variable is an array
+	r, ok := p0.Variable.Accept(t).(*types.DataType)
+	if !ok {
+		panic("expected data type")
+	}
+
+	if !r.IsArray {
+		panic("expected array")
+	}
+
+	return &types.DataType{
+		Name: r.Name,
+	}
+}
+
+func (t *typingVisitor) VisitStatementProcedureCall(p0 *parser.StatementProcedureCall) any {
+	returns := t.analyzeProcedureCall(p0.Call)
+
+	// a procedure that returns data can choose
+	// not to use the return values
+	if len(p0.Variables) == 0 {
+		return nil
+	}
+
+	if len(returns.Fields) != len(p0.Variables) {
+		panic(fmt.Sprintf("expected %d return values, got %d", len(returns.Fields), len(p0.Variables)))
+	}
+
+	if returns.IsTable {
+		panic("procedure returns a table, which cannot be assigned to variables")
+	}
+
+	for i, v := range p0.Variables {
+		if v == nil {
+			// skip if nil, since it is an anonymous receiver
+			t.anonymousReceiverTypes = append(t.anonymousReceiverTypes, returns.Fields[i].Type)
+			continue
+		}
+		varType, ok := t.declarations[*v]
+		if !ok {
+			panic(fmt.Sprintf("variable %s not declared", reverseCleanVar(*v)))
+		}
+
+		if !varType.Equals(returns.Fields[i].Type) {
+			panic(fmt.Sprintf("variable %s has wrong type", reverseCleanVar(*v)))
+		}
+	}
+
+	return nil
+}
+
+// TODO: this is a hack and implicit coupling with the clean package
+func reverseCleanVar(v string) string {
+	if !strings.HasPrefix(v, "_param_") {
+		panic("expected parameter prefix, received: " + v)
+	}
+
+	return "$" + strings.TrimPrefix(v, "_param_")
+}
```

Comment:
I guess this is not going to be called normally, but it jumped out that it does `HasPrefix` twice in the positive path.  This would be more straightforward and efficient:

```go
func reverseCleanVar(v string) string {
	v, prefixed := strings.CutPrefix(v, "_param_")
	if !prefixed {
		panic("expected parameter prefix, received: " + v) // already in a panic, just do nothing and return v
	}
	return "$" + v
}
```

But if it's only ever going to be used inside another `panic`, might as well just return something and have the outer panic work using that regardless of if the prefix was there.